src/main/java/org/openrewrite/java/migrate/javax/HttpSessionInvalidate.java

```diff
@@ -0,0 +1,102 @@
+    }
+
+    @Override
+    public String getDescription() {
+        return "Updates `javax.servlet.http.HttpSession invalidate()` to `javax.servlet.http.HttpServletRequest logout()`";
+    }
+
+    private final MethodMatcher INVALIDATE_METHOD_PATTERN = new MethodMatcher("javax.servlet.http.HttpSession invalidate()", false);
+
+    @Override
+    public TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new HttpSessionVisitor();
+    }
+
+    public class HttpSessionVisitor extends JavaIsoVisitor<ExecutionContext> {
+        @Override
+        public J.MethodInvocation visitMethodInvocation(J.MethodInvocation method, ExecutionContext ctx) {
+            if (INVALIDATE_METHOD_PATTERN.matches(method)) {
+                // Get index of param for HttpServletRequest, from the encapsulating method declaration TODO: would like to make this cleaner...
+                J.MethodDeclaration parentMethod = getCursor().dropParentUntil(parent -> parent instanceof J.MethodDeclaration).getValue();
+                Integer servletReqParamIndex = getServletRequestIndex(parentMethod);
+
+                // failed to find HttpServletRequest from parent MethodDeclaration
+                if (servletReqParamIndex == null) {
+                    return method;
+                }
+
+                // Get the HttpServletRequest param
+                J.VariableDeclarations httpServletRequestDeclaration = (J.VariableDeclarations) parentMethod.getParameters().get(servletReqParamIndex);
+
+                // Replace HttpSession.invalidate() with HttpServletRequest.logout()
+                final JavaTemplate logoutTemplate =
+                        JavaTemplate.builder("#{any(javax.servlet.http.HttpServletRequest)}.logout()")
+                                .imports("javax.servlet.http.HttpServletRequest")
+                                .javaParser(JavaParser.fromJavaVersion().classpathFromResources(ctx, "javax.servlet-3.0"))
+                                .build();
+                maybeAddImport("javax.servlet.http.HttpServletRequest");
+                method = logoutTemplate.apply(
+                        getCursor(),
+                        method.getCoordinates().replace(),
+                        httpServletRequestDeclaration.getVariables().get(0)
+                );
+                return method;
+            }
+            return method;
+        }
+
+        /**
+         * @return the param index position of the HttpServletRequest parameter object
+         */
+        @Nullable
+        private Integer getServletRequestIndex(J.MethodDeclaration parentMethod) {
+            List<JavaType> params = parentMethod.getMethodType().getParameterTypes();
+            for (int i = 0; i < params.size(); ++i) {
+                String paramType = ((JavaType.Class) params.get(i)).getFullyQualifiedName();
+                if (paramType.equals("javax.servlet.http.HttpServletRequest")) {
```

Comment:
We prefer to match using `org.openrewrite.java.TypeMatcher` instead of doing String comparisons. That then also handles any missing types correctly.
```suggestion
                if (HTTP_SERVLET_REQUEST_TYPE_MATCHER.matches(params.get(i))) {
```
With the addition of a field
```java
final TypeMatcher HTTP_SERVLET_REQUEST_TYPE_MATCHER = new TypeMatcher("javax.servlet.http.HttpServletRequest");
```